launchpadstation/launchpadstation.ino

```diff
@@ -0,0 +1,296 @@
+// Pins where the relays are connected
+#define PIN_RELAY_FIRE A2 // Write LOW to this pin to enable the ignition circuit
+#define PIN_RELAY_FILL A3 // Write LOW to this pin to open solenoid 1
+#define PIN_RELAY_VENT A4 // Write LOW to this pin to open solenoid 2
+
+#define CMD_FILL_START 0x61 // 'a'
+#define CMD_FILL_STOP  0x62 // 'b'
+#define CMD_VENT_START 0x63 // 'c'
+#define CMD_VENT_STOP  0x64 // 'd'
+#define CMD_ARM        0x65 // 'e'
+#define CMD_DISARM     0x66 // 'f'
+#define CMD_FIRE_START 0x67 // 'g'
+#define CMD_FIRE_STOP  0x68 // 'h'
+#define CMD_TM_ENABLE  0x41 // 'A'
+#define CMD_TM_DISABLE 0x42 // 'B'
+#define CMD_CA_TRIGGER 0x43 // 'C'
+
+#define REPLY_ACK  0x3D // '=' This is returned if the command is successfuly executed
+#define REPLY_NACK 0x21 // '!' This is returned if the command is not successfuly executed
+
+uint8_t command = 0x00;
+bool is_filling = false;
+bool is_venting = false;
+bool is_armed = false;
+
+void setup()
+{
+  pinMode(PIN_RELAY_FIRE, OUTPUT);
+  pinMode(PIN_RELAY_FILL, OUTPUT);
+  pinMode(PIN_RELAY_VENT, OUTPUT);
+  pinMode(PIN_OMBI_TM, OUTPUT);
+  pinMode(PIN_OMBI_CA, OUTPUT);
+  // Disable the ignition circuit
+  digitalWrite(PIN_RELAY_FIRE, HIGH);
+  // Close the solenoids
+  digitalWrite(PIN_RELAY_FILL, HIGH);
+  digitalWrite(PIN_RELAY_VENT, HIGH);
+  // Default state is HIGH for the ombilicals
+  digitalWrite(PIN_OMBI_TM, HIGH);
+  digitalWrite(PIN_OMBI_CA, HIGH);
+
+  init_communication();
+}
+
+void loop()
+{
+  read_byte(&command);
+
+  if (command)
+  {
+    switch (command)
+    {
+    case CMD_FILL_START:
+      start_filling();
+      break;
+    case CMD_FILL_STOP:
+      stop_filling();
+      break;
+    case CMD_VENT_START:
+      start_venting();
+      break;
+    case CMD_VENT_STOP:
+      stop_venting();
+      break;
+    case CMD_ARM:
+      arm();
+      break;
+    case CMD_DISARM:
+      disarm();
+      break;
+    case CMD_FIRE_START:
+      start_ignition();
+      break;
+    case CMD_FIRE_STOP:
+      stop_ignition();
+      break;
+    case CMD_TM_ENABLE:
+      enable_telemetry();
+      break;
+    case CMD_TM_DISABLE:
+      disable_telemetry();
+      break;
+    case CMD_CA_TRIGGER:
+      trigger_calibration();
+      break;
+    default:
+    send_byte(REPLY_NACK);
+      break;
+    }
+  }
+  // Reset this to the default value
+  command = 0x00;
+  delay(100);
+}
+
+/*
+ * Functions for communication with the Ground Station
+ */
+
+void init_communication()
+{ // Initialize the communication link
+  Serial.begin(115200);
+}
+
+void read_byte(uint8_t *data)
+{ // Read one byte in the buffer
+  if (Serial.available() > 0)
+  {
+    *data = Serial.read();
+  }
+}
+
+void send_byte(uint8_t data)
+{ // Write one byte to the communication link
+  Serial.write(data);
+}
+
+/*
+ * Controls for the Rocket fueling and ignition
+ */
+
+void start_filling()
+{ // Enable solenoid 1 only if solenoid 2 is disabled
+  if (!is_venting && !is_armed)
```

Comment:
Change if(..)'s if we are not gonna use it. We have space to have 4 different buttons in the GUI for fill, vent, arm and ingite.

Reply:
Or maybe armed is on always after ingite only, so we dont fill or something after ignition?